app/utils/JourneyChecker.scala

```diff
@@ -79,31 +71,47 @@ trait JourneyChecker extends IncomeSourcesUtils {
       }
     }
 
+  private lazy val homeRedirectUrl: (JourneyType) => MtdItUser[_] => Future[Result] =
+    (journeyType: JourneyType) => user => {
+      Future.successful {
+        (journeyType.operation, isAgent(user)) match {
+          case (Add, false) => Redirect(controllers.incomeSources.add.routes.AddIncomeSourceController.show())
+          case (Add, true) => Redirect(controllers.incomeSources.add.routes.AddIncomeSourceController.showAgent())
+          case (Manage, _) => Redirect(controllers.incomeSources.manage.routes.ManageIncomeSourceController.show(isAgent(user)))
+          case (Cease, false) => Redirect(controllers.incomeSources.cease.routes.CeaseIncomeSourceController.show())
+          case (Cease, true) => Redirect(controllers.incomeSources.cease.routes.CeaseIncomeSourceController.showAgent())
+        }
+      }
+    }
+
   private def useDefaultRedirect(data: UIJourneySessionData, journeyType: JourneyType, midwayFlag: Boolean): Boolean = {
     journeyType.operation match {
       case Add => !(midwayFlag && data.addIncomeSourceData.flatMap(_.incomeSourceAdded).getOrElse(false))
       case _ => true
     }
   }
 
-  def withSessionData(journeyType: JourneyType, midwayFlag: Boolean = true)(codeBlock: UIJourneySessionData => Future[Result])
+  def withSessionData(journeyType: JourneyType, midwayFlag: Boolean = true, initialPage: Boolean = false, nonLoopFlag: Boolean = false)(codeBlock: UIJourneySessionData => Future[Result])
```

Comment:
midwayFlag and initialPage representing state where user at the moment within the actual workflow, these can be combined into a single enumeration:

```
   State | midwayFlag | initialPage | WorkflowState
   A     |   false    | false   | None
   B     |   true     | false   | MidWay
   C     |   false    | true    | InitialPage
   D     |   true     | true    | Not possible scenario
```

Reply:
I would suggest to avoid using defaults in this method, its fine when we resolving compile time errors;